File: src/parser/parser.cpp
/* Parser - Thalia syntax analyzer
 * Copyright (C) 2021 Stan Vlad <vstan02@protonmail.com>
 *
 * This file is part of Thalia.
 *
 * Thalia is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <utility>

#include "parser/parser.hpp"
#include "parser/exception.hpp"

using namespace thalia::parser;

extern std::vector<stmts::statement*> parser::parse() {
	if (_tokens.empty()) {
		_tokens = _target.scan();
	}
	
	if (_tokens.size() < 2) {
		return _ast;
	}

	_index = 1;

	if (_ast.empty()) {
		while (!at_end()) {
			_ast.push_back(declaration());
		}
	}
	return _ast;
}

extern stmts::statement* parser::declaration() {
	using lexer::token_type;
	switch (_tokens[_index].type) {
		case token_type::VAR:
			return var_declaration();
		default:
			return program_declaration();
	}
}

extern stmts::statement* parser::var_declaration() {
	advance();
	lexer::token token = _tokens[_index];
	advance();
	consume(lexer::token_type::SEMICOLON, "Expect ';' after value.");
	return new stmts::var(token);
}

extern stmts::statement* parser::program_declaration() {
	consume(lexer::token_type::PROGRAM, "Expect 'program'.");
	if (!check(lexer::token_type::LEFT_BRACE)) {
		throw exception("Expect '{' after 'program'.", _tokens[_index].line);
	}
	return new stmts::program(block_statement());
}

extern stmts::statement* parser::print_statement() {
	advance();
	std::vector<exprs::expression*> target;
	target.push_back(expression());
	while(check(lexer::token_type::COMMA)) {
		advance();
		target.push_back(expression());
	}
	consume(lexer::token_type::SEMICOLON, "Expect ';' after value.");
	return new stmts::print(target);
}

extern stmts::statement* parser::block_statement() {
	advance();
	std::vector<stmts::statement*> stmts;
	while (!check(lexer::token_type::RIGHT_BRACE)) {
		stmts.push_back(statement());
	}
	consume(lexer::token_type::RIGHT_BRACE, "Expect '}' after block.");
	return new stmts::block(stmts);
}

extern stmts::statement* parser::if_statement() {
	advance();
	exprs::expression* condition = expression();
	if (!check(lexer::token_type::LEFT_BRACE)) {
		throw exception("Expect '{' after 'if'.", _tokens[_index].line);
	}
	return new stmts::if_(condition, block_statement());
}

extern stmts::statement* parser::while_statement() {
	advance();
	exprs::expression* condition = expression();
	if (!check(lexer::token_type::LEFT_BRACE)) {
		throw exception("Expect '{' after 'while'.", _tokens[_index].line);
	}
	return new stmts::while_(condition, block_statement());
}

extern stmts::statement* parser::each_statement() {
	using lexer::token_type;
	advance();
	consume(token_type::IDENTIFIER, "Expect a variable name after 'each'.");
	exprs::variable* variable = new exprs::variable(_tokens[_index - 1]);
	
	consume(token_type::IN, "Expect the keyword 'in' after the variable in 'each'.");
	consume(token_type::LEFT_BRACKET, "Expect '[' after 'in'.");
	exprs::expression* to = expression();

	exprs::expression* from = nullptr;
	if (check(token_type::COMMA)) {
		advance();
		from = expression();
		std::swap(from, to);
	}
	
	consume(token_type::RIGHT_BRACKET, "Expect ']' after 'each' interval.");
	return new stmts::each(variable, from, to, block_statement());
}

extern stmts::statement* parser::expression_statement() {
	exprs::expression* target = expression();
	consume(lexer::token_type::SEMICOLON, "Expect ';' after value.");
	return new stmts::expression(target);
}

extern stmts::statement* parser::statement() {
	using lexer::token_type;
	switch (_tokens[_index].type) {
		case token_type::PRINT:
			return print_statement();
		case token_type::LEFT_BRACE:
			return block_statement();
		case token_type::IF:
			return if_statement();
		case token_type::WHILE:
			return while_statement();
		case token_type::EACH:
			return each_statement();
		default:
			return expression_statement();
	}
}

extern exprs::expression* parser::assign_expression() {
	using lexer::token_type;
	exprs::expression* result = equality_expression();

	if (!match({
		token_type::EQUAL,
		token_type::PLUS_EQUAL, 
		token_type::MINUS_EQUAL,
		token_type::SLASH_EQUAL,
		token_type::STAR_EQUAL,
		token_type::PERCENT_EQUAL 
	})) {
		return result;
	}

	if (result->type != exprs::expr_type::VARIABLE) {
		throw exception("Invalid assignment target.", _tokens[_index - 2].line);
	}

	lexer::token token = _tokens[_index - 1];
	if (token.type == token_type::EQUAL) {
		return new exprs::assign(result, expression());
	}

	token.type = to_operation(token.type);
	exprs::expression* value = new exprs::binary(token, result, expression());
	exprs::expression* variable = new exprs::variable((static_cast<exprs::variable*>(result))->name);
	return new exprs::assign(variable, value);
}

extern exprs::expression* parser::equality_expression() {
	using lexer::token_type;
	exprs::expression* result = comparison_expression();
	while (match({ token_type::EQUAL_EQUAL, token_type::BANG_EQUAL })) {
		lexer::token token = _tokens[_index - 1];
		result = new exprs::binary(token, result, comparison_expression());
	}
	return result;
}

extern exprs::expression* parser::comparison_expression() {
	using lexer::token_type;
	exprs::expression* result = term_expression();
	while (match({ token_type::GREATER, token_type::GREATER_EQUAL, token_type::LESS, token_type::LESS_EQUAL })) {
		lexer::token token = _tokens[_index - 1];
		result = new exprs::binary(token, result, term_expression());
	}
	return result;
}

extern exprs::expression* parser::term_expression() {
	using lexer::token_type;
	exprs::expression* result = factor_expression();
	while (match({ token_type::PLUS, token_type::MINUS })) {
		lexer::token token = _tokens[_index - 1];
		result = new exprs::binary(token, result, factor_expression());
	}
	return result;
}

extern exprs::expression* parser::factor_expression() {
	using lexer::token_type;
	exprs::expression* result = unary_expression();
	while (match({ token_type::STAR, token_type::SLASH, token_type::PERCENT })) {
		lexer::token token = _tokens[_index - 1];
		result = new exprs::binary(token, result, unary_expression());
	}
	return result;
}

extern exprs::expression* parser::unary_expression() {
	using lexer::token_type;
	if (match({ token_type::BANG, token_type::MINUS })) {
		lexer::token token = _tokens[_index - 1];
		return new exprs::unary(token, unary_expression());
	}
	return primary_expression();
}

extern exprs::expression* parser::primary_expression() {
	using lexer::token_type;
	if (check(token_type::IDENTIFIER)) {
		advance();
		return new exprs::variable(_tokens[_index - 1]);
	}

	if (check(token_type::LEFT_PAREN)) {
		advance();
		exprs::expression* result = expression();
		consume(lexer::token_type::RIGHT_PAREN, "Expect ')' after expression.");
		return new exprs::grouping(result);
	}

	advance();
	return new exprs::literal(_tokens[_index - 1]);
}

extern bool parser::match(std::vector<lexer::token_type> types) {
	for (lexer::token_type type: types) {
		if (check(type)) {
			advance();
			return true;
		}
	}
	return false;
}

extern void parser::consume(lexer::token_type type, const char* message) {
	if (check(type)) {
		return advance();
	}
	throw exception(message, _tokens[_index].line);
}

extern thalia::lexer::token_type parser::to_operation(lexer::token_type type) {
	using lexer::token_type;
	switch (type) {
		case token_type::PLUS_EQUAL:	
			return token_type::PLUS;
		case token_type::MINUS_EQUAL:	
			return token_type::MINUS;
		case token_type::STAR_EQUAL:	
			return token_type::STAR;
		case token_type::SLASH_EQUAL:	
			return token_type::SLASH;
		case token_type::PERCENT_EQUAL:	
			return token_type::PERCENT;
		default:
			return token_type::END;
	}
}

File: src/thalia/parser.clj
;; Copyright (C) 2024 Stan Vlad <vstan02@protonmail.com>
;;
;; This file is part of Thalia.
;;
;; Thalia is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <https://www.gnu.org/licenses/>.

(ns thalia.parser
  (:gen-class)
  (:require [thalia.token :as token]))

(def ^:private parse-expr nil)
(def ^:private parse-stmt nil)

(defn ^:private at-end? [tokens]
  (or (token/check (first tokens) #{:EOF}) (empty? tokens)))

(defn ^:private skip-until [types tokens]
  (drop-while #(not (token/check % types)) tokens))

(defn ^:private gen-parse-repeated [delims parse-item]
  (fn [tokens]
    (let [$res1 (parse-item tokens)]
      (loop [$nodes [(:ast $res1)]
             $errors (:errors $res1)
             $tokens (:rest $res1)]
        (if-not (token/check (first $tokens) delims)
          {:ast $nodes :errors $errors :rest $tokens}
          (let [$res2 (->> $tokens rest parse-item)]
            (recur (conj $nodes (:ast $res2))
                   (concat $errors (:errors $res2))
                   (:rest $res2))))))))

(defn ^:private parse-pattern [nname tokens rules]
  (loop [node {:type nname}
         errors []
         rules0 rules
         tokens0 tokens]
    (if (or (empty? rules0) (empty? tokens0))
      {:ast node :errors errors :rest tokens0}
      (let [rule (first rules0)
            tkn (first tokens0)]
        (if (some? (:values rule))
          (if (token/check tkn (:values rule))
            (recur (merge (when (:field rule) {(:field rule) tkn}) node)
                   errors
                   (rest rules0)
                   (rest tokens0))
            {:ast node
             :errors (conj errors {:type (:error rule) :token tkn})
             :rest tokens0})
          (let [res ((:function rule) tokens0)]
            (recur (merge {(:field rule) (:ast res)} node)
                   (concat errors (:errors res))
                   (rest rules0)
                   (:rest res))))))))

(defn ^:private parse-expr-binary [tokens types parse-operand]
  (let [res1 (parse-operand tokens)]
    (loop [node (:ast res1)
           errors (:errors res1)
           tokens0 (:rest res1)]
      (if-not (token/check (first tokens0) types)
        {:ast node :errors errors :rest tokens0}
        (let [res2 (parse-pattern
                    :EXPR-BINARY tokens0
                    [{:values types :error :EXPECT-BINARY-OPERATOR :field :operator}
                     {:function parse-operand :field :right}])]
          (recur (merge {:left node} (:ast res2))
                 (concat errors (:errors res2))
                 (:rest res2)))))))

(defn ^:private parse-each-values [tokens]
  (let [value1 (parse-expr tokens)
        tokens1 (:rest value1)]
    (if (token/check (first tokens1) #{:COLON})
      (let [value2 (->> tokens1 rest parse-expr)
            tokens2 (:rest value2)]
        (if (token/check (first tokens2) #{:COLON})
          (let [value3 (->> tokens2 rest parse-expr)]
            {:ast {:from (:ast value1) :to (:ast value3) :step (:ast value2)}
             :errors (concat (:errors value1) (:errors value2) (:errors value3))
             :rest (:rest value3)})
          {:ast {:from (:ast value1) :to (:ast value2)}
           :errors (concat (:errors value1) (:errors value2))
           :rest (:rest value2)}))
      {:ast {:to (:ast value1)} :errors (:errors value1) :rest tokens1})))

(defn ^:private parse-expr-primary [tokens]
  (case (->> tokens first :type)
    :LPAREN
    (parse-pattern
     :EXPR-GROUPING tokens
     [{:values #{:LPAREN} :error :EXPECT-LPAREN}
      {:function parse-expr :field :value}
      {:values #{:RPAREN} :error :EXPECT-RPAREN}])
    :ID
    (parse-pattern
     :EXPR-VARIABLE tokens
     [{:values #{:ID} :error :EXPECT-VARIABLE :field :token}])
    (parse-pattern
     :EXPR-LITERAL tokens
     [{:values #{:INT} :error :EXPECT-LITERAL :field :token}])))

(defn ^:private parse-expr-unary [tokens]
  (if-not (token/check (first tokens) #{:BANG :MINUS})
    (parse-expr-primary tokens)
    (parse-pattern
     :EXPR-UNARY tokens
     [{:values #{:MINUS :BANG} :error :EXPECT-UNARY-OPERATOR :field :operator}
      {:function parse-expr-unary :field :value}])))

(defn ^:private parse-expr-factor [tokens]
  (parse-expr-binary
   tokens
   #{:STAR :SLASH :PERCENT}
   parse-expr-unary))

(defn ^:private parse-expr-term [tokens]
  (parse-expr-binary
   tokens
   #{:PLUS :MINUS}
   parse-expr-factor))

(defn ^:private parse-expr-comp [tokens]
  (parse-expr-binary
   tokens
   #{:GREATER :LESS :GREATER-EQUAL :LESS-EQUAL}
   parse-expr-term))

(defn ^:private parse-expr-equal [tokens]
  (parse-expr-binary
   tokens
   #{:EQUAL-EQUAL :BANG-EQUAL}
   parse-expr-comp))

(defn ^:private parse-expr-logic-and [tokens]
  (parse-expr-binary
   tokens
   #{:AMPER-AMPER}
   parse-expr-equal))

(defn ^:private parse-expr-logic-or [tokens]
  (parse-expr-binary
   tokens
   #{:PIPE-PIPE}
   parse-expr-logic-and))

(defn ^:private parse-expr-assign [tokens]
  (let [res1 (parse-expr-logic-or tokens)]
    (if-not (token/check (->> res1 :rest first) #{:EQUAL})
      res1
      (let [res2 (->> res1 :rest rest parse-expr)]
        {:ast {:type :EXPR-ASSIGN
               :target (:ast res1)
               :value (:ast res2)}
         :errors (concat (:errors res1) (:errors res2))
         :rest (:rest res2)}))))

(defn ^:private parse-expr [tokens]
  (parse-expr-assign tokens))

(defn ^:private parse-stmt-expression [tokens]
  (parse-pattern
   :STMT-EXPRESSION tokens
   [{:function parse-expr :field :value}
    {:values #{:SEMI} :error :EXPECT-SEMI}]))

(defn ^:private parse-stmt-print [tokens]
  (->> [{:values #{:PRINT} :error :EXPECT-PRINT}
        (when-not (token/check (second tokens) #{:SEMI})
          {:function (gen-parse-repeated #{:COMMA} parse-expr) :field :values})
        {:values #{:SEMI} :error :EXPECT-SEMI}]
       (filter some?)
       (parse-pattern :STMT-PRINT tokens)
       ((fn [res] {:rest (:rest res)
                   :errors (:errors res)
                   :ast (merge {:values []} (:ast res))}))))

(defn ^:private parse-stmt-println [tokens]
  (->> [{:values #{:PRINTLN} :error :EXPECT-PRINTLN}
        (when-not (token/check (second tokens) #{:SEMI})
          {:function (gen-parse-repeated #{:COMMA} parse-expr) :field :values})
        {:values #{:SEMI} :error :EXPECT-SEMI}]
       (filter some?)
       (parse-pattern :STMT-PRINTLN tokens)
       ((fn [res] {:rest (:rest res)
                   :errors (:errors res)
                   :ast (merge {:values []} (:ast res))}))))

(defn ^:private parse-stmt-block [tokens]
  (if-not (token/check (first tokens) #{:LBRACE})
    {:ast {:type :STMT-BLOCK}
     :errors {:type :EXPECT-LBRACE :token (first tokens)}
     :rest tokens}
    (loop [nodes []
           errors []
           tokens0 (rest tokens)]
      (cond
        (token/check (first tokens0) #{:RBRACE})
        {:ast {:type :STMT-BLOCK :stmts nodes}
         :errors errors
         :rest (rest tokens0)}
        (token/check (first tokens0) #{:EOF})
        {:ast {:type :STMT-BLOCK}
         :errors [{:type :EXPECT-RBRACE :token (first tokens0)}]
         :rest tokens0}
        :else
        (let [res (parse-stmt tokens0)]
          (recur (conj nodes (:ast res))
                 (concat errors (:errors res))
                 (:rest res)))))))

(defn ^:private parse-stmt-if [tokens]
  (let [value1 (parse-pattern
                :STMT-IF tokens
                [{:values #{:IF} :error :EXPECT-IF}
                 {:function parse-expr :field :condition}
                 {:function parse-stmt-block :field :body}])
        tokens1 (:rest value1)]
    (if-not (token/check (first tokens1) #{:ELSE})
      value1
      (let [tokens2 (rest tokens1)
            value2 (if (token/check (first tokens2) #{:IF})
                     (parse-stmt-if tokens2)
                     (parse-stmt-block tokens2))]
        {:ast (merge (:ast value1) {:else (:ast value2)})
         :errors (concat (:errors value1) (:errors value2))
         :rest (:rest value2)}))))

(defn ^:private parse-stmt-while [tokens]
  (parse-pattern
   :STMT-WHILE tokens
   [{:values #{:WHILE} :error :EXPECT-WHILE}
    {:function parse-expr :field :condition}
    {:function parse-stmt-block :field :body}]))

(defn ^:private parse-stmt-each [tokens]
  (parse-pattern
   :STMT-EACH tokens
   [{:values #{:EACH} :error :EXPECT-EACH}
    {:values #{:ID} :error :EXPECT-ID :field :target}
    {:values #{:IN} :error :EXPECT-IN}
    {:values #{:LBRACKET} :error :EXPECT-LBRACKET}
    {:function parse-each-values :field :values}
    {:values #{:RBRACKET} :error :EXPECT-RBRACKET}
    {:function parse-stmt-block :field :body}]))

(defn ^:private parse-stmt [tokens]
  (case (->> tokens first :type)
    :PRINT (parse-stmt-print tokens)
    :PRINTLN (parse-stmt-println tokens)
    :LBRACE (parse-stmt-block tokens)
    :IF (parse-stmt-if tokens)
    :WHILE (parse-stmt-while tokens)
    :EACH (parse-stmt-each tokens)
    (parse-stmt-expression tokens)))

(defn ^:private parse-decl-variable [tokens]
  (parse-pattern
   :DECL-VARIABLE tokens
   [{:values #{:VAR} :error :EXPECT-VAR}
    {:values #{:ID} :error :EXPECT-ID :field :token}
    {:values #{:SEMI} :error :EXPECT-SEMI}]))

(defn ^:private parse-decl-program [tokens]
  (parse-pattern
   :DECL-PROGRAM tokens
   [{:values #{:PROGRAM} :error :EXPECT-PROGRAM}
    {:function parse-stmt-block :field :body}]))

(defn parse-decl [tokens]
  (case (->> tokens first :type)
    :VAR (parse-decl-variable tokens)
    :PROGRAM (parse-decl-program tokens)
    {:ast {}
     :errors [{:type :INVAL-DECL :token (first tokens)}]
     :rest (skip-until #{:VAR :PROGRAM} tokens)}))

(defn parse [tokens]
  (loop [nodes []
         errors []
         tokens0 tokens]
    (if (at-end? tokens0)
      {:ast nodes :errors errors :rest tokens0}
      (let [res (parse-decl tokens0)]
        (recur (conj nodes (:ast res))
               (concat errors (:errors res))
               (:rest res))))))

